### PR TITLE
Backport ValueProvider from [BEAM-551]

### PR DESCRIPTION
**Pull request by sammcveety**
R: @davorbonaci

### CODE REVIEW

**Comment by davorbonaci** [comment]
R: @dhalperi, for more context

**Comment by dhalperi** [comment]
For reviewing purposes, which Beam PR is this a backport of? Need to be able to compare diffs 1-1 to confirm backport success.

**Comment by sammcveety** [comment]
https://github.com/apache/incubator-beam/pull/852
https://github.com/apache/incubator-beam/pull/1009
https://github.com/apache/incubator-beam/pull/1105
https://github.com/apache/incubator-beam/pull/1107
https://github.com/apache/incubator-beam/pull/1146
https://github.com/apache/incubator-beam/pull/1266

**Comment by sammcveety** [comment]
retest this please

**Comment by davorbonaci** [comment]
"retest this please" doesn't work in Dataflow, only Beam. I kicked the build through the Travis UI.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/options/ProxyInvocationHandlerTest.java [hunk 1]
```diff
@@ -245,27 +245,31 @@ public void testToString() throws Exception {
   public void testToStringAfterDeserializationContainsJsonEntries() throws Exception {
```
Comment (dhalperi) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/options/ProxyInvocationHandlerTest.java]: Good: https://www.diffchecker.com/jL9V73Wf

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/options/ValueProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright (C) 2016 Google Inc.
+ *
```
Comment (dhalperi) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/options/ValueProviderTest.java]: Missing test snippet: https://www.diffchecker.com/Hbe3kotU

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/options/ValueProviderUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
+/*
```
Comment (dhalperi) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/options/ValueProviderUtilsTest.java]: Lovely https://www.diffchecker.com/vjqGG2B7

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataTest.java [hunk 4]
```diff
@@ -39,6 +39,8 @@
 import static org.junit.Assert.assertThat;
```
Comment (dhalperi) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataTest.java]: diff seems fine: https://www.diffchecker.com/60YVy0SU

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/options/ProxyInvocationHandler.java [hunk 5]
```diff
@@ -495,6 +523,43 @@ private Object getValueFromJson(String propertyName, Method method) {
   }
 
   /**
+   * Helper method to return standard Default cases.
```
Comment (sammcveety) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/options/ProxyInvocationHandler.java]: Done

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/options/ValueProviderTest.java (hunk 2 above)
Comment (sammcveety) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/options/ValueProviderTest.java]: Done

Comment (dhalperi) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/options/ValueProviderTest.java]: https://www.diffchecker.com/XwXfeyaw -- can you keep the spacing consistent?

**Comment by sammcveety** [comment]
Done

**Comment by dhalperi** [comment]
LGTM: https://www.diffchecker.com/FsAidkqL

Will merge once green.

**Pull request merged**